exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,561 @@
+    fn validator_id(snapshot: &dyn Snapshot, pub_key: &PublicKey) -> Option<ValidatorId> {
+        CoreSchema::new(snapshot)
+            .consensus_config()
+            .find_validator(|validator_keys| *pub_key == validator_keys.service_key)
+    }
+
+    /// If the current node is a validator, return its identifier, for other nodes return `default`.
+    fn validator_id_or(
+        snapshot: &dyn Snapshot,
+        pub_key: &PublicKey,
+        default: i32
+    ) -> i32 {
+        Self::validator_id(snapshot, pub_key)
+            .map_or(default, |id| i32::from(id.0))
+    }
+
+    /// Converts `JniResult<R>` into `Result<R, ExecutionError>`.
+    fn convert_jni<R>(res: JniResult<R>) -> Result<R, ExecutionError> {
+        res.map_err(|err| {
+            let kind: ErrorKind = match err.kind() {
+                JniErrorKind::JavaException => Error::JavaException.into(),
+                _ => Error::OtherJniError.into(),
+            };
+            (kind, err).into()
+        })
+    }
+
+    /// Handles and clears any Java exceptions.
+    ///
+    /// Any JNI errors are converted into `ExecutionError` with their descriptions, for JNI errors
+    /// like `JniErrorKind::JavaException` it gets (and clears) any exception that is currently
+    /// being thrown, then exception is passed to corresponding `ExceptionHandler` according their
+    /// type and `exception_handlers` mapping.
+    /// `ExceptionHandlers::DEFAULT` is called in case of there is no any handlers or handlers are
+    /// not matched to exception type.
+    fn handle_exception<H, R>(
+        env: &JNIEnv,
+        result: JniResult<R>,
+        exception_handlers: &[(&GlobalRef, H)],
+    ) -> Result<R, ExecutionError>
+        where
+            H: Fn(&JNIEnv, JObject) -> ExecutionError,
+    {
+        result.map_err(|err| {
+            match err.kind() {
+                JniErrorKind::JavaException => {
+                    let exception = get_and_clear_java_exception(env);
+                    for (class, handler) in exception_handlers {
+                        if unwrap_jni(env.is_instance_of(exception, *class)) {
+                            return handler(env, exception);
+                        }
+                    }
+
+                    ExceptionHandlers::DEFAULT(env, exception)
+                },
+                _ => (Error::OtherJniError, err).into(),
+            }
+        })
+    }
+
+    /// Handles and clears any Java exceptions.
+    ///
+    /// Any JNI errors are converted into `ExecutionError` with their descriptions, for JNI errors
+    /// like `JniErrorKind::JavaException` it gets (and clears) any exception that is currently
+    /// being thrown, then exception is passed to `ExceptionHandlers::DEFAULT`.
+    fn handle_any_exception<R>(
+        env: &JNIEnv,
+        result: JniResult<R>,
+    ) -> Result<R, ExecutionError>
+    {
+        result.map_err(|err| {
```

Comment:
redundant code duplication, it's enough to call `handle_exception` with empty array